# Disable OSGi metadata generation and remove bnd annotations.
# Tomcat doesn't use OSGi at runtime, almost nobody deploys Tomcat as OSGi
# bundles, and the OSGi ecosystem has largely been superseded by Java 9+
# modules (JPMS). Not worth packaging bnd/osgi just for unused manifest
# entries.
diff -urN apache-tomcat-11.0.27-src.orig/build.xml apache-tomcat-11.0.27-src/build.xml
--- apache-tomcat-11.0.27-src.orig/build.xml	2026-10-06 20:12:07.000000000 +0200
+++ apache-tomcat-11.0.27-src/build.xml	2026-10-28 11:29:26.199908248 +0100
@@ -1140,7 +1140,7 @@
       filesDir="${tomcat.classes}"
       filesId="files.annotations-api"
       manifest="${tomcat.manifests}/annotations-api.jar.manifest"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- Servlet Implementation JAR File -->
     <jarIt jarfile="${servlet-api.jar}"
@@ -1149,48 +1149,48 @@
       manifest="${tomcat.manifests}/servlet-api.jar.manifest"
       notice="${tomcat.manifests}/servlet-api.jar.notice"
       license="${tomcat.manifests}/servlet-api.jar.license"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- EL Implementation JAR File -->
     <jarIt jarfile="${el-api.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.el-api"
       manifest="${tomcat.manifests}/el-api.jar.manifest"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- JSP Implementation JAR File -->
     <jarIt jarfile="${jsp-api.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.jsp-api"
       manifest="${tomcat.manifests}/jsp-api.jar.manifest"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- WebSocket API JAR File -->
     <jarIt jarfile="${websocket-api.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.websocket-api"
       manifest="${tomcat.manifests}/websocket-api.jar.manifest"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- WebSocket Client API JAR File -->
     <jarIt jarfile="${websocket-client-api.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.websocket-client-api"
       manifest="${tomcat.manifests}/websocket-client-api.jar.manifest"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- JASPIC API JAR File -->
     <jarIt jarfile="${jaspic-api.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.jaspic-api"
       manifest="${tomcat.manifests}/jaspic-api.jar.manifest"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- Tomcat-juli JAR File -->
     <jarIt jarfile="${tomcat-juli.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.tomcat-juli"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- Bootstrap JAR File -->
     <jarIt jarfile="${bootstrap.jar}"
@@ -1202,68 +1202,68 @@
     <jarIt jarfile="${tomcat-util.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.tomcat-util"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- Tomcat API JAR File -->
     <jarIt jarfile="${tomcat-api.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.tomcat-api"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- Tomcat Util Scan JAR File -->
     <jarIt jarfile="${tomcat-util-scan.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.tomcat-util-scan"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <jarIt jarfile="${tomcat-jni.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.tomcat-jni"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- Protocol handlers - Coyote -->
     <jarIt jarfile="${tomcat-coyote.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.tomcat-coyote"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- OpenSSL FFM - Coyote -->
     <jarIt jarfile="${tomcat-coyote-ffm.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.tomcat-coyote-ffm"
       manifest="${tomcat.manifests}/tomcat-coyote-ffm.jar.manifest"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- WebSocket implementation JAR File -->
     <jarIt jarfile="${tomcat-websocket.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.tomcat-websocket"
       meta-inf="${tomcat.manifests}/tomcat-websocket.jar"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- Catalina GroupCom/Tribes JAR File -->
     <jarIt jarfile="${catalina-tribes.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.catalina-tribes"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- Catalina Main JAR File -->
     <jarIt jarfile="${catalina.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.catalina"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- Catalina Cluster/HA JAR File -->
     <jarIt jarfile="${catalina-ha.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.catalina-ha"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- Server-Side Includes (SSI) -->
     <jarIt jarfile="${catalina-ssi.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.catalina-ssi"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- Catalina Ant Tasks JAR File -->
     <jarIt jarfile="${catalina-ant.jar}"
@@ -1274,27 +1274,27 @@
     <jarIt jarfile="${catalina-storeconfig.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.catalina-storeconfig"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- Jasper EL Implementation JAR File -->
     <jarIt jarfile="${jasper-el.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.jasper-el"
       meta-inf="${tomcat.manifests}/jasper-el.jar"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- Jasper Implementation JAR File -->
     <jarIt jarfile="${jasper.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.jasper"
       meta-inf="${tomcat.manifests}/jasper.jar"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- Re-packaged Apache Commons DBCP 2-->
     <jarIt jarfile="${tomcat-dbcp.jar}"
       filesDir="${tomcat.classes}"
       filesId="files.tomcat-dbcp"
-      addOSGi="true" />
+      addOSGi="false" />
 
     <!-- i18n JARs -->
     <jar jarfile="${tomcat.build}/lib/tomcat-i18n-cs.jar"
@@ -1758,7 +1758,7 @@
            filesId="files.tomcat-embed-core"
            notice="${tomcat.manifests}/servlet-api.jar.notice"
            license="${tomcat.manifests}/servlet-api.jar.license"
-           addOSGi="true"
+           addOSGi="false"
            addGraal="true"
            graalPrefix="org.apache.tomcat.embed/tomcat-embed-core"
            graalFiles="res/graal/tomcat-embed-core/native-image"
@@ -1766,7 +1766,7 @@
     <jarIt jarfile="${tomcat-embed-el.jar}"
            filesDir="${tomcat.classes}"
            filesId="files.tomcat-embed-el"
-           addOSGi="true"
+           addOSGi="false"
            addGraal="true"
            graalPrefix="org.apache.tomcat.embed/tomcat-embed-el"
            graalFiles="res/graal/tomcat-embed-el/native-image"
@@ -1775,7 +1775,7 @@
            filesDir="${tomcat.classes}"
            filesId="files.tomcat-embed-jasper"
            meta-inf="${tomcat.manifests}/jasper.jar"
-           addOSGi="true"
+           addOSGi="false"
            addGraal="true"
            graalPrefix="org.apache.tomcat.embed/tomcat-embed-jasper"
            graalFiles="res/graal/tomcat-embed-jasper/native-image"
@@ -1784,7 +1784,7 @@
            filesDir="${tomcat.classes}"
            filesId="files.tomcat-embed-websocket"
            meta-inf="${tomcat.manifests}/tomcat-websocket.jar"
-           addOSGi="true"
+           addOSGi="false"
            addGraal="true"
            graalPrefix="org.apache.tomcat.embed/tomcat-embed-websocket"
            graalFiles="res/graal/tomcat-embed-websocket/native-image"
@@ -4350,26 +4350,9 @@
     <taskdef name="jsign" classname="net.jsign.JsignTask" classpathref="jsign.classpath" />
   </target>
 
-  <target name="download-bnd" >
-    <!-- Download bnd -->
-    <antcall target="downloadfile">
-      <param name="sourcefile" value="${bnd.loc}"/>
-      <param name="destfile" value="${bnd.jar}"/>
-      <param name="destdir" value="${bnd.home}"/>
-      <param name="checksum.enabled" value="${bnd.checksum.enabled}"/>
-      <param name="checksum.algorithm" value="${bnd.checksum.algorithm}"/>
-      <param name="checksum.value" value="${bnd.checksum.value}"/>
-    </antcall>
-  </target>
-
-  <target name="setup-bnd" depends="download-bnd" unless="${skip.build.java.version}">
-    <!-- Add bnd tasks to project -->
-    <path id="bnd.classpath">
-      <fileset file="${bnd.jar}" />
-    </path>
+  <target name="download-bnd" />
 
-    <taskdef resource="aQute/bnd/ant/taskdef.properties" classpathref="bnd.classpath" />
-  </target>
+  <target name="setup-bnd" depends="download-bnd" />
 
   <macrodef name="jarIt" description="utility macro for standard JAR packaging">
     <attribute name="jarfile"
diff -urN apache-tomcat-11.0.27-src.orig/java/org/apache/el/ExpressionFactoryImpl.java apache-tomcat-11.0.27-src/java/org/apache/el/ExpressionFactoryImpl.java
--- apache-tomcat-11.0.27-src.orig/java/org/apache/el/ExpressionFactoryImpl.java	2026-10-06 20:12:07.000000000 +0200
+++ apache-tomcat-11.0.27-src/java/org/apache/el/ExpressionFactoryImpl.java	2026-10-28 11:29:26.205409680 +0100
@@ -31,7 +31,6 @@
 /**
  * Implementation of {@link jakarta.el.ExpressionFactory}.
  */
-@aQute.bnd.annotation.spi.ServiceProvider(value = ExpressionFactory.class)
 public class ExpressionFactoryImpl extends ExpressionFactory {
 
     /**
diff -urN apache-tomcat-11.0.27-src.orig/java/org/apache/juli/logging/LogFactory.java apache-tomcat-11.0.27-src/java/org/apache/juli/logging/LogFactory.java
--- apache-tomcat-11.0.27-src.orig/java/org/apache/juli/logging/LogFactory.java	2026-10-06 20:12:07.000000000 +0200
+++ apache-tomcat-11.0.27-src/java/org/apache/juli/logging/LogFactory.java	2026-10-28 11:29:26.200650431 +0100
@@ -21,7 +21,6 @@
 import java.util.ServiceLoader;
 import java.util.logging.LogManager;
 
-import aQute.bnd.annotation.spi.ServiceConsumer;
 
 /**
  * This is a modified LogFactory that uses a simple {@link ServiceLoader} based discovery mechanism with a default of
@@ -48,7 +47,6 @@
  * <strong>IMPLEMENTATION NOTE</strong> - This implementation is heavily based on the SAXParserFactory and
  * DocumentBuilderFactory implementations (corresponding to the JAXP pluggability APIs) found in Apache Xerces.
  */
-@ServiceConsumer(value = Log.class)
 public class LogFactory {
 
     private static final LogFactory singleton = new LogFactory();
diff -urN apache-tomcat-11.0.27-src.orig/java/org/apache/tomcat/websocket/server/DefaultServerEndpointConfigurator.java apache-tomcat-11.0.27-src/java/org/apache/tomcat/websocket/server/DefaultServerEndpointConfigurator.java
--- apache-tomcat-11.0.27-src.orig/java/org/apache/tomcat/websocket/server/DefaultServerEndpointConfigurator.java	2026-10-06 20:12:07.000000000 +0200
+++ apache-tomcat-11.0.27-src/java/org/apache/tomcat/websocket/server/DefaultServerEndpointConfigurator.java	2026-10-28 11:29:26.209305886 +0100
@@ -30,7 +30,6 @@
  * Default implementation of {@link ServerEndpointConfig.Configurator} for WebSocket server endpoints.
  */
 @SuppressWarnings("javadoc") // Eclipse bug - the Javadoc is fine
-@aQute.bnd.annotation.spi.ServiceProvider(value = ServerEndpointConfig.Configurator.class)
 public class DefaultServerEndpointConfigurator extends ServerEndpointConfig.Configurator {
 
     /**
diff -urN apache-tomcat-11.0.27-src.orig/java/org/apache/tomcat/websocket/WsContainerProvider.java apache-tomcat-11.0.27-src/java/org/apache/tomcat/websocket/WsContainerProvider.java
--- apache-tomcat-11.0.27-src.orig/java/org/apache/tomcat/websocket/WsContainerProvider.java	2026-10-06 20:12:07.000000000 +0200
+++ apache-tomcat-11.0.27-src/java/org/apache/tomcat/websocket/WsContainerProvider.java	2026-10-28 11:29:26.212771169 +0100
@@ -22,7 +22,6 @@
 /**
  * WebSocket container provider for Tomcat.
  */
-@aQute.bnd.annotation.spi.ServiceProvider(value = ContainerProvider.class)
 public class WsContainerProvider extends ContainerProvider {
 
     /**
